src/nvim/fileio.c

```diff
@@ -7154,7 +7169,26 @@ char_u *getnextac(int c, void *cookie, int indent)
     acp->nextcmd = ac->next;
   }
 
-  return retval;
+  if (retval != NULL) {
+    // Remove one-shot ("once") autocmd in anticipation of its execution.
+    if (ac->once) {
+      XFREE_CLEAR(ac->cmd);
+      au_need_clean = true;
+    }
+    return retval;
+  } else {
+    assert(ac->lua_cmd != LUA_NOREF);
+    Array args = ARRAY_DICT_INIT;
+    executor_exec_lua_cb(ac->lua_cmd, "autocommand", args, false);
+    if (ac->once) {
```

Comment:
That's slightly different semantics here — with VimL we mark command as removed before executing but here we are marking it as removed after we execute the callback.

I think I'm confused how cleanup of aucmds works now.

Reply:
It could make difference with both ++once and ++nested at the same time. I would suggest to "take" the reference, by saving `ac->lua_cmd` to a local varable and (when `++once` ) setting `ac->lua_cmd = LUA_NOREF` before calling.

Reply:
Done, I've left code which frees lua ref in `au_cleanup` for "just in case" but my understanding is that it should free it at the moment of marking as deleted.